ethereum_history_api/circuits/lib/src/merkle_patricia_proofs/proof.nr

```diff
@@ -23,6 +25,45 @@ pub fn leaf_get_len<MAX_LEN>(leaf: Leaf<MAX_LEN>) -> u32 {
     leaf_unpadded_len as u32
 }
 
+// Prefixes mean:
+// 0 - Extension node, even number of nibbles
+// 1 - Extension node, odd number of nibbles
+// 2 - Leaf node, even number of nibbles
+// 3 - Leaf node, odd number of nibbles
+pub fn parity(prefix: u8) -> u64 {
+    assert([0, 1, 2, 3].any(|x| x == prefix), "Prefix must be 0, 1, 2 or 3");
+    let mut result = -1;
+    if (prefix % 2 == 0) {
+        result = EVEN;
+    } else {
+        result = ODD;
+    }
+    assert((result == EVEN) | (result == ODD), "Invalid parity");
+
+    result
+}
+
+// Leaf and extension nodes are RLP lists with 2 elements: prefixed key nibbles (in bytes) and value.
+// This function removes the prefix and returns appropriate number of key nibbles according to the parity of the prefix (see function: parity).
+pub fn unprefixed_key_nibbles<MAX_KEY_LEN, MAX_KEY_NIBBLE_LEN>(prefixed_key_bytes: [u8; MAX_KEY_LEN]) -> [u8; MAX_KEY_NIBBLE_LEN] {
+    assert(MAX_KEY_LEN * 2 == MAX_KEY_NIBBLE_LEN);
+
+    let prefixed_key_nibbles: [u8; MAX_KEY_NIBBLE_LEN] = bytes_to_nibbles(prefixed_key_bytes);
+    let mut unprefixed_key_nibbles = [0; MAX_KEY_NIBBLE_LEN];
+
+    if (parity(prefixed_key_nibbles[0]) == ODD) {
+        memcpy_up_to_length(&mut unprefixed_key_nibbles, prefixed_key_nibbles, 1, MAX_KEY_NIBBLE_LEN - 1);
```

Comment:
Simpler implementation:
```rust
let parity = parity(prefixed_key_nibbles[0]);
let offset = 1 + parity;
let length = MAX_KEY_NIBBLE_LEN - 1 - parity;
memcpy_up_to_length(&mut unprefixed_key_nibbles, prefixed_key_nibbles, offset, length);
```

Reply:
With such a simple implementation, perhaps parity could be inlined (not insisting thou)

Reply:
I would leave parity as a function because 1) it has comment explanation what parity is, 2) it has assertion that parity < 4 that I think belongs there

Reply:
I implemented your version, slightly different, you can see if it's better now